### PR TITLE
[mlir][transform] Add support for transform.param pad multiples in `PadOp`

### PR DESCRIPTION
**Pull request by srcarroll**
This patch modifies the definition of `PadOp` to take transform params and handles for the `pad_to_multiple_of` operand.

### CODE REVIEW

**Comment by srcarroll** [comment]
This is kinda messy, but I'm not familiar enough with some of the utilities to find a good way to reduce common code.  Suggestions are very welcome

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td [hunk 1]
```diff
@@ -1021,8 +1023,7 @@ def PadOp : Op<Transform_Dialect, "structured.pad",
                       TransformHandleTypeInterface:$pad,
                       TransformHandleTypeInterface:$copy);
 
-  let assemblyFormat =
-    "$target attr-dict `:` functional-type(operands, results)";
+  let hasCustomAssemblyFormat = 1;
```
Comment (MaheshRavishankar) [inline mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: Do we really need a custom assembly format. The IR changes dont seem to need this.

**Review by MaheshRavishankar — DISMISSED** [review]

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td (hunk 1 above)
Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: I couldn't figure out how to avoid it.  I was following what was done for `TileUsingForOp` (which I also did for `PadOp` in a previous patch).  I welcome suggestions.

Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: actually i remember now. using the available directives would result in having the param/handle types inlined into the list they are passed to, e.g.
```
    %1:2 = transform.structured.tile_using_forall %0 tile_sizes [%sz : !transform.any_op, 20]
           : (!transform.any_op) -> (!transform.any_op, !transform.any_op)
```
opposed to
```
    %1:2 = transform.structured.tile_using_forall %0 tile_sizes [%sz, 20]
           : (!transform.any_op, !transform.any_op) -> (!transform.any_op, !transform.any_op)
```
like how `tile_using_for` does it. i prefered the latter and followed it for both the pad and vectorize op

Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: I actually wouldn't be opposed to using the dynamic list directive in favor of simpler code. even if I do personally think the resulting IR is uglier, but obviously will put personal preferences aside. Would like to hear what others think.

Comment (MaheshRavishankar) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: I dont know about the transform ops, but the ops that implement the `ViewLikeInterface` use the custom parser/printer hooks for handling this (https://github.com/llvm/llvm-project/blob/b86e0992bfa6c58be077d82d824016f590ac5d90/mlir/include/mlir/Interfaces/ViewLikeInterface.h#L107) . Maybe that could be refactored a little bit to allow using that without needing the op to implement the interface. AFAIK, those dont have this problem.

Comment (ftynse) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: > I actually wouldn't be opposed to using the dynamic list directive for this in favor of simpler code. even if I do personally think the resulting IR is uglier, but obviously will put personal preferences aside. But I think there should be consistency. So I would also suggest changing tile_using_for and vectorize if possible. Would like to hear what others think.

I strongly support consistency across transform ops. Going out of our way to add syntactic consistency with only tangentially-related interface is a waste of cycles. Take a look at https://github.com/llvm/llvm-project/blob/7dfb6f571c8090803a32b5611e519b77ae74de30/mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td#L2004-L2008, if _that_ can be reused here, let's do that. Otherwise, let's stick with custom syntax and eventually upgrade all syntax everywhere.

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td [hunk 2]
```diff
@@ -1043,11 +1050,12 @@ def PadOp : Op<Transform_Dialect, "structured.pad",
     /// copy_back_op attribute value indicating that no copy back is desired.
     static constexpr StringRef kCopyOpNone = "none";
 
-    ::mlir::DiagnosedSilenceableFailure applyToOne(
-        ::mlir::transform::TransformRewriter &rewriter,
-        ::mlir::linalg::LinalgOp target,
-        ::mlir::transform::ApplyToEachResultList &results,
-        ::mlir::transform::TransformState &state);
+    SmallVector<OpFoldResult> getMixedPadToMultipleOf();
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1053]: Nit: please add the `///` comment for this function.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 3]
```diff
@@ -171,6 +171,50 @@ static DiagnosedSilenceableFailure unpackSingleIndexResultPayloadOperations(
   return DiagnosedSilenceableFailure::success();
 }
 
+static DiagnosedSilenceableFailure reifyMixedParamAndHandleResults(
+    TransformState &state, TransformOpInterface &transformOp,
+    const SmallVectorImpl<OpFoldResult> &mixedResults,
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:176]: Nit: if you don't modify a vector, pass in an `ArrayRef` instead.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 4]
```diff
@@ -171,6 +171,50 @@ static DiagnosedSilenceableFailure unpackSingleIndexResultPayloadOperations(
   return DiagnosedSilenceableFailure::success();
 }
 
+static DiagnosedSilenceableFailure reifyMixedParamAndHandleResults(
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:174]: Nit: please document top-level functions.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 5]
```diff
@@ -171,6 +171,50 @@ static DiagnosedSilenceableFailure unpackSingleIndexResultPayloadOperations(
...
+    TransformState &state, TransformOpInterface &transformOp,
+    const SmallVectorImpl<OpFoldResult> &mixedResults,
+    SmallVectorImpl<int64_t> &reified) {
+  for (OpFoldResult paramOrHandle : mixedResults) {
+    if (isa<Attribute>(paramOrHandle)) {
+      reified.push_back(
+          cast<IntegerAttr>(paramOrHandle.get<Attribute>()).getInt());
+      continue;
+    } else if (isa<Value>(paramOrHandle) &&
+               isa<ParamType>(paramOrHandle.get<Value>().getType())) {
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:184]: Nit: at this point, we know that `paramOrHandle` is a `Value` so no need to check.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 6]
```diff
@@ -171,6 +171,50 @@ static DiagnosedSilenceableFailure unpackSingleIndexResultPayloadOperations(
...
+      ArrayRef<Attribute> params = state.getParams(paramOrHandle.get<Value>());
+      if (params.size() != 1)
+        return transformOp.emitDefiniteFailure() << "expected a single param";
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:187]: I'd consider a silenceable error here and below, as along as we haven't modified the payload.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 7]
```diff
@@ -171,6 +171,50 @@ static DiagnosedSilenceableFailure unpackSingleIndexResultPayloadOperations(
...
+      reified.push_back(
+          cast<IntegerAttr>(params.front()).getValue().getSExtValue());
+      continue;
+    }
+
+    auto paramOrHandlePayloads =
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:193]: Nit: just `payload`? At this point, we know it's not a param.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 8]
```diff
@@ -171,6 +171,50 @@ static DiagnosedSilenceableFailure unpackSingleIndexResultPayloadOperations(
...
+        state.getPayloadOps(paramOrHandle.get<Value>());
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:194]: On the other hand, we haven't checked that it isn't a value handle.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 9]
```diff
@@ -1677,18 +1723,126 @@ void transform::PadOp::build(OpBuilder &b, OperationState &result, Value target,
...
+  consumesHandle(getTarget(), effects);
+  onlyReadsHandle(getPadToMultipleOf(), effects);
+  producesHandle(getPadded(), effects);
+  producesHandle(getPad(), effects);
+  producesHandle(getCopy(), effects);
+  modifiesPayload(effects);
+}
+
+SmallVector<OpFoldResult> PadOp::getMixedPadToMultipleOf() {
+  OpBuilder b(getContext());
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:1771]: It is preferable to take a reference to `OpBuilder`. This lets the caller configure notification hooks on the builder and not be surprised by new operations appearing out of thin air. However, in this case, you don't need an `OpBuilder`, but just a `Builder`, for which it is okay to have a local instance.

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td (hunk 1 above)
Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: yah that's the one i was talking about. maybe i was using it wrong. let me try again

File: mlir/python/mlir/dialects/transform/structured.py [hunk 10]
```diff
@@ -373,10 +373,11 @@ class PadOp(PadOp):
     def __init__(
         self,
         target: Union[Operation, OpView, Value],
+        pad_to_multiple_of: Optional[Union[DynamicIndexList, ArrayAttr]] = None,
         *,
         padding_values: Optional[Union[ArrayAttr, Sequence[Attribute]]] = None,
         padding_dimensions: OptionalIntList = None,
-        pad_to_multiple_of: OptionalIntList = None,
+        static_pad_to_multiple_of: OptionalIntList = None,
```
Comment (ftynse) [inline mlir/python/mlir/dialects/transform/structured.py:380]: Can we rather take this as another option for `pad_to_multiple_of` and then dispatch based on type?

File: mlir/python/mlir/dialects/transform/structured.py [hunk 11]
```diff
@@ -373,10 +373,11 @@ class PadOp(PadOp):
     def __init__(
         self,
         target: Union[Operation, OpView, Value],
+        pad_to_multiple_of: Optional[Union[DynamicIndexList, ArrayAttr]] = None,
```
Comment (ftynse) [inline mlir/python/mlir/dialects/transform/structured.py:376]: I'm not sure what happens with typing here. Ideally, we should take a list where one can mix python integers, handles and attributes freely OR an array attribute.

File: mlir/python/mlir/dialects/transform/structured.py
Comment (ftynse) [inline mlir/python/mlir/dialects/transform/structured.py:1]: Do we also need to update the vectorize op, or is it already in this form?

**Review by ftynse — COMMENTED** [review]
Looks okay to me. Please address some minor comments.

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td (hunk 1 above)
Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: oh, so i'm an idiot and was just using it wrong. fixed now. thanks!

Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: i would like to do the same for the other ops. i'll submit another pr for that

Comment (ftynse) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1026]: Thanks!

File: mlir/python/mlir/dialects/transform/structured.py
Comment (srcarroll) [reply mlir/python/mlir/dialects/transform/structured.py:1]: it was already in that form. my vectorize pr only enabled params, handles were already there.

File: mlir/python/mlir/dialects/transform/structured.py (hunk 11 above)
Comment (srcarroll) [reply mlir/python/mlir/dialects/transform/structured.py:376]: to be honest, i pretty much copy pasta'd from vectorize op for this. i might need to think about this to understand

File: mlir/python/mlir/dialects/transform/structured.py (hunk 10 above)
Comment (srcarroll) [reply mlir/python/mlir/dialects/transform/structured.py:380]: not exactly sure what you mean

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp (hunk 9 above)
Comment (srcarroll) [reply mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:1771]: i see. i was just going with the same signature i've seen in similar functions. i'll switch to `Builder`

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp (hunk 7 above)
Comment (srcarroll) [reply mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:193]: true

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp (hunk 8 above)
Comment (srcarroll) [reply mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:194]: hmm, not quite sure what you mean

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td (hunk 1 above)
Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td]: so now `vectorize`, `tile_using_for` and `pad_op` have the same syntax, where the dynamic types are in the functional-type. the `tile_using_forall` op still has the types inlined in the dynamic list. so i guess we should decide, for consistency, which one to go with. and i'll go ahead and make a PR to update all of them.

Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td]: although it looks like `PackedOrDynamicIndexList` doesn't let the types be optional like `DynamicIndexList` does

Comment (srcarroll) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td]: fyi, i started updating to make consistent. but i think `PackedOrDynamicIndexList` will need modifiying in order to get `tile_using_forall` to work the same. https://github.com/srcarroll/llvm-project/compare/pad-op-parameter...srcarroll:llvm-project:consistent-transform-syntax

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td [hunk 12]
```diff
@@ -1021,8 +1023,13 @@ def PadOp : Op<Transform_Dialect, "structured.pad",
                       TransformHandleTypeInterface:$pad,
                       TransformHandleTypeInterface:$copy);
 
-  let assemblyFormat =
-    "$target attr-dict `:` functional-type(operands, results)";
+  let assemblyFormat = [{
+    $target oilist(
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:1027]: You don't need `oilist` unless there is a list (separated by `|`) inside it.

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp (hunk 8 above)
Comment (ftynse) [reply mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp]: We have (operation) handles and value handles. The fact the something is not a parameter doesn't necessarily make it an operation handle that `getPayloadOps` expects, it may be a value handle. So this will have to check, and react accordingly, if `paramOrHandle` is a value handle. Or we can disallow value handles in the type constraint of the op that uses this and add an assertion here.

File: mlir/python/mlir/dialects/transform/structured.py (hunk 11 above)
Comment (ftynse) [reply mlir/python/mlir/dialects/transform/structured.py:376]: Just check out the definition of `DynamicIndexList`. It seems to be more or less what I described.

File: mlir/python/mlir/dialects/transform/structured.py (hunk 10 above)
Comment (ftynse) [reply mlir/python/mlir/dialects/transform/structured.py:380]: I don't think we should take a separate `static_pad_to_multiple_of` in Python, we don't do that in C++ builders and take mixed values instead. This exposes unnecessary implementation detail of the operation. In fact, we can already provide a list of python integers into `pad_to_multiple_of`, we just need to forward that correctly in the body.

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td (hunk 1 above)
Comment (ftynse) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td]: I'm fine changing things to get more consistency.

File: mlir/python/mlir/dialects/transform/structured.py (hunk 10 above)
Comment (srcarroll) [reply mlir/python/mlir/dialects/transform/structured.py:380]: makes sense. will do

Comment (srcarroll) [reply mlir/python/mlir/dialects/transform/structured.py]: done. please check

File: mlir/python/mlir/dialects/transform/structured.py (hunk 11 above)
Comment (srcarroll) [reply mlir/python/mlir/dialects/transform/structured.py]: yes. so i believe this typing is correct then?

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp (hunk 8 above)
Comment (srcarroll) [reply mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp]: so i guess i was confused by your other comment https://github.com/llvm/llvm-project/pull/90755#discussion_r1587706648. so i do indeed need to check if value for both then?

File: mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp [hunk 13]
```diff
@@ -171,6 +171,51 @@ static DiagnosedSilenceableFailure unpackSingleIndexResultPayloadOperations(
...
+    } else if (isa<ParamType>(paramOrHandle.get<Value>().getType())) {
+      ArrayRef<Attribute> params = state.getParams(paramOrHandle.get<Value>());
+      if (params.size() != 1)
+        return transformOp.emitSilenceableError() << "expected a single param";
+      reified.push_back(
+          cast<IntegerAttr>(params.front()).getValue().getSExtValue());
+      continue;
+    }
+
+    auto payload = state.getPayloadOps(paramOrHandle.get<Value>());
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp:195]:
```suggestion
    Value handle = paramOrHandle.get<Value>();
    if (!isa<TransformHandleOpInterface>(handle.getType())
      return transformOp.emitSilenceableError() << "unexpected value handle";
    auto payload = state.getPayloadOps(handle);
```

value handles are accepted by the syntax, but not supported below. Calling `getPayloadOps` on a value handle will assert.

**Review by ftynse — APPROVED** [review]
